src/transformers/trainer.py

```diff
@@ -584,8 +584,10 @@ def _remove_unused_columns(self, dataset: "datasets.Dataset", description: Optio
         else:
             return dataset.remove_columns(ignored_columns)
 
-    def _get_train_sampler(self) -> Optional[torch.utils.data.Sampler]:
-        if not has_length(self.train_dataset):
+    def _get_train_sampler(self, train_dataset: Optional[Dataset] = None) -> Optional[torch.utils.data.Sampler]:
+        train_dataset = train_dataset if train_dataset is not None else self.train_dataset
+
+        if not has_length(train_dataset):
```

Comment:
Let's make it explicit that we return `None` here when `self.train_dataset is None`.

```suggestion
        if self.train_dataset is None or not has_length(self.train_dataset):
```